src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
+        loop {
+            let mut new_ts = (self.now)();
+            new_ts -= new_ts % self.update_interval_ms;
+            if self.upper.less_than(&new_ts) {
+                break new_ts;
+            }
+            let sleep_ms = self.update_interval_ms - (new_ts % self.update_interval_ms);
+            tokio::time::sleep(Duration::from_millis(sleep_ms)).await;
         }
     }
+
+    /// Returns an iterator of timestamp bindings for a given partition
+    fn partition_uppers(&self, pid: &PartitionId) -> PartitionUppers {
+        let bindings = match self.remap_trace.get(pid) {
+            Some(bindings) => (*bindings).iter(),
+            None => (&[]).iter(),
+        };
+        PartitionUppers {
+            offset: MzOffset::default(),
+            bindings,
+        }
+    }
+}
+
+/// The Iterator returned by [ReclockOperator::partition_uppers]
+struct PartitionUppers<'a> {
+    offset: MzOffset,
+    bindings: std::slice::Iter<'a, (Timestamp, MzOffset)>,
+}
+
+impl Iterator for PartitionUppers<'_> {
+    type Item = (Timestamp, MzOffset);
+    fn next(&mut self) -> Option<Self::Item> {
+        let &(ts, diff) = self.bindings.next()?;
+        self.offset += diff;
+        Some((ts, self.offset))
+    }
+}
+
+/// The Iterator returned by [ReclockOperator::reclock]
+pub struct ReclockIter<'a, M> {
+    reclock: &'a ReclockOperator,
+    messages: hash_map::IterMut<'a, PartitionId, Vec<(M, MzOffset)>>,
+}
+
+impl<'a, M> Iterator for ReclockIter<'a, M> {
+    type Item = (PartitionId, ReclockPartIter<'a, M>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (partition, messages) = self.messages.next()?;
+        Some((
+            partition.clone(),
+            ReclockPartIter {
+                bindings: self.reclock.partition_uppers(partition).peekable(),
+                messages: messages.drain(..),
+            },
+        ))
+    }
+}
+
+/// The Iterator returned by [ReclockIter::next]
+pub struct ReclockPartIter<'a, M> {
+    bindings: Peekable<PartitionUppers<'a>>,
+    messages: std::vec::Drain<'a, (M, MzOffset)>,
+}
+
+impl<'a, M> Iterator for ReclockPartIter<'a, M> {
+    type Item = (M, Timestamp);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (message, offset) = self.messages.next()?;
+        // Skip bindings whose source offset upper doesn't cover this message's offset
+        while !(offset < self.bindings.peek().expect("not enough bindings").1) {
+            self.bindings.next();
+        }
```

Comment:
note that this depends on `self.messages` being sorted

Reply:
Indeed, and they are sorted in the `reclock` method